### ISSUE TITLE
Troubleshoot instructions for CentOS 7

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#3222]
Reportedly the INSTALL.md instructions for CentOS 7 do not work out of the box. We need to verify that:

- [x] Confirm documentation is up-to-date. That is, following the instructions as written works.
- [x] Confirm all unit tests pass in a clean install
- [x] Confirm shared object libraries usable by test apps
- [x] Get SDK working with openssl 1.1.1 (1.0.2 is EOL this year), may require new instructions and/or updating the current instructions.

I am going to use this issue as a log of any problems I discover, may get noisy before I solve any problems.

**Comment by coryan** [comment]
To start, I wanted to verify the code works on CentOS.  I ran:

```console
# Copied from the INSTALL.md instructions.
sudo yum makecache && sudo yum install -y automake cmake3 curl-devel gcc gcc-c++ git libtool         make openssl-devel pkgconfig tar wget which zlib-devel
sudo ln -sf /usr/bin/cmake3 /usr/bin/cmake
```

Then I cloned my repository (in case I need to make changes and fixes)


```console
> git clone git@github.com:coryan/google-cloud-cpp.git cbt-develop
> cd cbt-develop
> git rev-parse --short HEAD
27a726c
```

And built the code with shared libraries using a super build:

```
cmake -Hsuper -Bcmake-out/super -DBUILD_SHARED_LIBS=yes
cmake --build cmake-out/super -- -j $(nproc)
```

That failed:

```
[ 76%] Performing build step for 'googleapis_project'
CMake Error at /home/coryan/cbt-develop/cmake-out/super/external/googleapis/src/googleapis_project-stamp/googleapis_project-build-.cmake:49 (message):
  Command failed: 2

   '/usr/bin/cmake3' '--build' '/home/coryan/cbt-develop/cmake-out/super/external/googleapis/src/googleapis_project-build' '--' '-j' '8'

  See also

    /home/coryan/cbt-develop/cmake-out/super/external/googleapis/src/googleapis_project-stamp/googleapis_project-build-*.log


gmake[2]: *** [external/googleapis/src/googleapis_project-stamp/googleapis_project-build] Error 1
gmake[1]: *** [CMakeFiles/googleapis_project.dir/all] Error 2
gmake: *** [all] Error 2
```

The actual error inside that is:

```
/home/coryan/cbt-develop/cmake-out/super/external/bin/grpc_cpp_plugin: error while loading shared libraries: libprotoc.so.3.7.1.0: cannot open shared object file: No such file or directory
--grpc_out: protoc-gen-grpc: Plugin failed with status code 127.
```

**Comment by coryan** [comment]
The problem is the gRPC plugin, it cannot find the shared libraries:

```console
> ldd /home/coryan/cbt-develop/cmake-out/super/external/bin/grpc_cpp_plugin
	linux-vdso.so.1 =>  (0x00007fff78692000)
	libprotoc.so.3.7.1.0 => not found
	libprotobuf.so.3.7.1.0 => not found
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fd359223000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fd35901b000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fd358dff000)
	libgrpc_plugin_support.so => not found
	libm.so.6 => /lib64/libm.so.6 (0x00007fd358afd000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fd3587f6000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fd3585e0000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fd358212000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd359427000)
```

That is because the program gets installed without a `RPATH` or `RUNPATH`:

```console
> objdump -p /home/coryan/cbt-develop/cmake-out/super/external/bin/grpc_cpp_plugin | egrep -c 'RPATH|RUNPATH'
0
```

Compare that to the `protoc` output:

```console
> objdump -p /home/coryan/cbt-develop/cmake-out/super/external/bin/protoc | egrep 'RPATH|RUNPATH'
  RPATH                $ORIGIN/../lib
```

That seems like a potential bug in gRPC, but we need to verify this is still a problem in the latest version.

**Comment by coryan** [comment]
Meanwhile we can complete the build using `LD_LIBRARY_PATH`, when these libraries get installed in `/usr/local/*` the `RPATH`/`RUNPATH` issues won't be a problem.  That works :tada: 

```console
> LD_LIBRARY_PATH=$PWD/cmake-out/super/external/lib cmake --build cmake-out/super -- -j $(nproc)
[snip snip snip]
139/140 Test #139: setup_test ................................................   Passed    0.03 sec
        Start 140: field_path_test
140/140 Test #140: field_path_test ...........................................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 140

Total Test time (real) =   9.95 sec
[100%] Completed 'google_cloud_cpp_project'
[100%] Built target google_cloud_cpp_project
```

**Comment by coryan** [comment]
Also our own binaries have the right `RPATH` set:

```console
> objdump -p cmake-out/super/build/google-cloud-cpp/src/google_cloud_cpp_project-build/google/cloud/bigtable/table_test | grep RPATH
  RPATH                /home/coryan/cbt-develop/cmake-out/super/external/lib64:/home/coryan/cbt-develop/cmake-out/super/external/lib
```

```console
> ldd cmake-out/super/build/google-cloud-cpp/src/google_cloud_cpp_project-build/google/cloud/bigtable/table_test | grep -c not_found
0
```

**Comment by coryan** [comment]
In summary:

- The code builds and runs with shared objects.
- At least gRPC, but maybe other dependencies, need `LD_LIBRARY_PATH` to build the code, but not to run.
  - Presumably installing them on one of the paths configured in `/etc/ld.so.conf` also works.

**Comment by coryan** [comment]
Next is to execute the steps in the INSTALL.md file, and fix any problems with it.  I did a couple of iterations before starting this log, and fixed some problems:

The instructions assume this directory exists:

```bash
mkdir -p $HOME/Downloads
```

```bash
sudo rpm -Uvh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
sudo yum install -y centos-release-scl
sudo yum-config-manager --enable rhel-server-rhscl-7-rpms
sudo yum makecache && \
sudo yum install -y automake cmake3 curl-devel gcc gcc-c++ git libtool \
        make openssl-devel pkgconfig tar wget which zlib-devel
sudo ln -sf /usr/bin/cmake3 /usr/bin/cmake && sudo ln -sf /usr/bin/ctest3 /usr/bin/ctest
```

```console
(echo "/usr/local/lib" ; echo "/usr/local/lib64") | \
sudo tee /etc/ld.so.conf.d/usrlocal.conf
```

```bash
cd $HOME/Downloads
wget -q https://github.com/google/crc32c/archive/1.0.6.tar.gz
tar -xf 1.0.6.tar.gz
cd $HOME/Downloads/crc32c-1.0.6
cmake \
      -DCMAKE_BUILD_TYPE=Release \
      -DBUILD_SHARED_LIBS=yes \
      -DCRC32C_BUILD_TESTS=OFF \
      -DCRC32C_BUILD_BENCHMARKS=OFF \
      -DCRC32C_USE_GLOG=OFF \
      -H. -Bcmake-out/crc32c
sudo cmake --build cmake-out/crc32c --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/google/protobuf/archive/v3.6.1.tar.gz
tar -xf v3.6.1.tar.gz
cd $HOME/Downloads/protobuf-3.6.1/cmake
cmake \
        -DCMAKE_BUILD_TYPE=Release \
        -DBUILD_SHARED_LIBS=yes \
        -Dprotobuf_BUILD_TESTS=OFF \
        -H. -Bcmake-out
sudo cmake --build cmake-out --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/c-ares/c-ares/archive/cares-1_14_0.tar.gz
tar -xf cares-1_14_0.tar.gz
cd $HOME/Downloads/c-ares-cares-1_14_0
./buildconf && ./configure && make -j ${NCPU:-4}
sudo make install
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/grpc/grpc/archive/v1.19.1.tar.gz
tar -xf v1.19.1.tar.gz
cd $HOME/Downloads/grpc-1.19.1
export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:/usr/local/lib64/pkgconfig
export PATH=/usr/local/bin:${PATH}
make -j ${NCPU:-4}
sudo make install
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/googleapis/cpp-cmakefiles/archive/v0.1.5.tar.gz
tar -xf v0.1.5.tar.gz
cd $HOME/Downloads/cpp-cmakefiles-0.1.5
cmake \
    -DBUILD_SHARED_LIBS=YES \
    -H. -Bcmake-out
sudo cmake --build cmake-out --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/google/googletest/archive/release-1.10.0.tar.gz
tar -xf release-1.10.0.tar.gz
cd $HOME/Downloads/googletest-release-1.10.0
cmake \
      -DCMAKE_BUILD_TYPE="Release" \
      -DBUILD_SHARED_LIBS=yes \
      -H. -Bcmake-out
cmake --build cmake-out -- -j ${NCPU:-4}
sudo cmake --build cmake-out --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp-common/archive/v0.13.0.tar.gz
tar -xf v0.13.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-common-0.13.0
cmake -H. -Bcmake-out \
    -DBUILD_TESTING=OFF \
    -DGOOGLE_CLOUD_CPP_TESTING_UTIL_ENABLE_INSTALL=ON
cmake --build cmake-out -- -j $(nproc)
sudo cmake --build cmake-out --target install
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp/archive/v0.14.0.tar.gz
tar -xf v0.14.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-0.14.0
cmake -H. -Bcmake-out
cmake --build cmake-out -- -j ${NCPU:-4}
cd $HOME/Downloads/google-cloud-cpp/cmake-out
ctest --output-on-failure
sudo cmake --build . --target install
```

**Comment by coryan** [comment]
That builds correctly, and the tests pass:

```console
> cd cmake-out
> ctest --output-on-failure -j $(nproc)
[snip snip snip]
161/163 Test   #8: internal_compiler_info_test ...............................   Passed    0.00 sec
162/163 Test  #17: internal_env_test .........................................   Passed    0.00 sec
163/163 Test  #92: client_test ...............................................   Passed    1.66 sec

100% tests passed, 0 tests failed out of 163

Total Test time (real) =   1.67 sec
```

**Comment by coryan** [comment]
I created a 
[scripted version](https://github.com/googleapis/google-cloud-cpp/files/3772767/build.sh.txt)
 of [the build step above](https://github.com/googleapis/google-cloud-cpp/issues/3222#issuecomment-546385915).

**Comment by coryan** [comment]
In summary:

* The build instructions will need some updates to work out of the box on CentOS, mostly small (but critical) changes to deal with shared library paths.
* With those changes (and using static libraries) the code builds and the tests pass.
  * I have not tested shared libraries, they probably work, but there may be other problems lurking here.

**Comment by coryan** [comment]
### Rebuild google-cloud-cpp-common with shared libraries

```
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp-common/archive/v0.13.0.tar.gz
tar -xf v0.13.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-common-0.13.0
cmake -H. -Bcmake-out \
    -DBUILD_SHARED_LIBS=yes \
    -DBUILD_TESTING=OFF \
    -DGOOGLE_CLOUD_CPP_TESTING_UTIL_ENABLE_INSTALL=ON
cmake --build cmake-out -- -j $(nproc)
sudo cmake --build cmake-out --target install
sudo ldconfig
```

### Rebuild google-cloud-cpp with shared libraries

```
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp/archive/v0.14.0.tar.gz
tar -xf v0.14.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-0.14.0
cmake -H. -Bcmake-out   -DBUILD_SHARED_LIBS=yes
cmake --build cmake-out -- -j ${NCPU:-4}
cd cmake-out
ctest --output-on-failure
sudo cmake --build . --target install
```

**Comment by coryan** [comment]
The tests passed with shared libraries too:

```console
> ctest
[snip snip snip]
162/163 Test #162: benchmark_utils_test ......................................   Passed    0.02 sec
        Start 163: grpc_error_delegate_test
163/163 Test #163: grpc_error_delegate_test ..................................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 163

Total Test time (real) =  12.26 sec

> ldd google/cloud/bigtable/table_test 
	linux-vdso.so.1 =>  (0x00007ffc363ca000)
	libbigtable_client_testing.so => /home/coryan/Downloads/google-cloud-cpp-0.14.0/cmake-out/google/cloud/bigtable/libbigtable_client_testing.so (0x00007f7bb09db000)
	libbigtable_client.so.1 => /home/coryan/Downloads/google-cloud-cpp-0.14.0/cmake-out/google/cloud/bigtable/libbigtable_client.so.1 (0x00007f7bb00ec000)
	libgoogle_cloud_cpp_testing.so => /home/coryan/Downloads/google-cloud-cpp-0.14.0/cmake-out/google/cloud/testing_util/libgoogle_cloud_cpp_testing.so (0x00007f7bafee0000)
[snip snip snip]

```

**Comment by coryan** [comment]
Used a [script](https://github.com/googleapis/google-cloud-cpp/files/3772943/build-shared.sh.txt)
 to repeat the installation with shared libraries.

Works fine too.

**Comment by coryan** [comment]
Downloaded OpenSSL 1.1.1:

```console
> cd $HOME/Downloads
> wget https://www.openssl.org/source/openssl-1.1.1d.tar.gz
> wget https://www.openssl.org/source/openssl-1.1.1d.tar.gz.sha256
> cat openssl-1.1.1d.tar.gz.sha256
1e3a91bc1f9dfce01af26026f856e064eab4c8ee0a8f457b5ae30b40b8b711f2
> diff openssl-1.1.1d.tar.gz.sha256 <(sha256sum openssl-1.1.1d.tar.gz | cut -f1 -d' ') || echo "Mismatch"
> tar -zxf openssl-1.1.1d.tar.gz
> cd openssl-1.1.1d
> ./config
> make -j $(nproc)
> sudo make install
```

**Comment by coryan** [comment]
### Create download directory

```bash
mkdir -p $HOME/Downloads
```

### Install dev tools


```bash
sudo rpm -Uvh https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
sudo yum install -y centos-release-scl yum-utils
sudo yum-config-manager --enable rhel-server-rhscl-7-rpms
sudo yum makecache && \
sudo yum install -y automake cmake3 curl-devel gcc gcc-c++ git libtool \
        make openssl-devel pkgconfig tar wget which zlib-devel
sudo ln -sf /usr/bin/cmake3 /usr/bin/cmake && sudo ln -sf /usr/bin/ctest3 /usr/bin/ctest
```

```console
(echo "/usr/local/lib" ; echo "/usr/local/lib64") | \
sudo tee /etc/ld.so.conf.d/usrlocal.conf
```

```bash
cd $HOME/Downloads
wget -q https://github.com/google/crc32c/archive/1.0.6.tar.gz
tar -xf 1.0.6.tar.gz
cd $HOME/Downloads/crc32c-1.0.6
cmake \
      -DCMAKE_BUILD_TYPE=Release \
      -DBUILD_SHARED_LIBS=yes \
      -DCRC32C_BUILD_TESTS=OFF \
      -DCRC32C_BUILD_BENCHMARKS=OFF \
      -DCRC32C_USE_GLOG=OFF \
      -H. -Bcmake-out/crc32c
sudo cmake --build cmake-out/crc32c --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/google/protobuf/archive/v3.6.1.tar.gz
tar -xf v3.6.1.tar.gz
cd $HOME/Downloads/protobuf-3.6.1/cmake
cmake \
        -DCMAKE_BUILD_TYPE=Release \
        -DBUILD_SHARED_LIBS=yes \
        -Dprotobuf_BUILD_TESTS=OFF \
        -H. -Bcmake-out
sudo cmake --build cmake-out --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/c-ares/c-ares/archive/cares-1_14_0.tar.gz
tar -xf cares-1_14_0.tar.gz
cd $HOME/Downloads/c-ares-cares-1_14_0
./buildconf && ./configure && make -j ${NCPU:-4}
sudo make install
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/grpc/grpc/archive/v1.19.1.tar.gz
tar -xf v1.19.1.tar.gz
cd $HOME/Downloads/grpc-1.19.1
export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig:/usr/local/lib64/pkgconfig
export PATH=/usr/local/bin:${PATH}
make -j ${NCPU:-4}
sudo make install
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/googleapis/cpp-cmakefiles/archive/v0.1.5.tar.gz
tar -xf v0.1.5.tar.gz
cd $HOME/Downloads/cpp-cmakefiles-0.1.5
cmake \
    -DBUILD_SHARED_LIBS=YES \
    -H. -Bcmake-out
sudo cmake --build cmake-out --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/google/googletest/archive/release-1.10.0.tar.gz
tar -xf release-1.10.0.tar.gz
cd $HOME/Downloads/googletest-release-1.10.0
cmake \
      -DCMAKE_BUILD_TYPE="Release" \
      -DBUILD_SHARED_LIBS=yes \
      -H. -Bcmake-out
cmake --build cmake-out -- -j ${NCPU:-4}
sudo cmake --build cmake-out --target install -- -j ${NCPU:-4}
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp-common/archive/v0.13.0.tar.gz
tar -xf v0.13.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-common-0.13.0
cmake -H. -Bcmake-out \
    -DBUILD_TESTING=OFF \
    -DGOOGLE_CLOUD_CPP_TESTING_UTIL_ENABLE_INSTALL=ON
cmake --build cmake-out -- -j $(nproc)
sudo cmake --build cmake-out --target install
sudo ldconfig
```

```bash
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp/archive/v0.14.0.tar.gz
tar -xf v0.14.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-0.14.0
cmake -H. -Bcmake-out
cmake --build cmake-out -- -j ${NCPU:-4}
cd cmake-out
ctest --output-on-failure
sudo cmake --build . --target install
```


### Rebuild google-cloud-cpp-common with shared libraries

```
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp-common/archive/v0.13.0.tar.gz
tar -xf v0.13.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-common-0.13.0
cmake -H. -Bcmake-out \
    -DBUILD_SHARED_LIBS=yes \
    -DBUILD_TESTING=OFF \
    -DGOOGLE_CLOUD_CPP_TESTING_UTIL_ENABLE_INSTALL=ON
cmake --build cmake-out -- -j $(nproc)
sudo cmake --build cmake-out --target install
sudo ldconfig
```

### Rebuild google-cloud-cpp with shared libraries

```
cd $HOME/Downloads
wget -q https://github.com/googleapis/google-cloud-cpp/archive/v0.14.0.tar.gz
tar -xf v0.14.0.tar.gz
cd $HOME/Downloads/google-cloud-cpp-0.14.0
cmake -H. -Bcmake-out   -DBUILD_SHARED_LIBS=yes
cmake --build cmake-out -- -j ${NCPU:-4}
cd cmake-out
ctest --output-on-failure
sudo cmake --build . --target install
```

**Comment by coryan** [comment]
The previous commands worked for me.  I turned them into a [script](https://github.com/googleapis/google-cloud-cpp/files/3773479/build-openssl-1.1.1.txt).

The tests passed too, as show [in the log](https://github.com/googleapis/google-cloud-cpp/files/3773481/build-openssl-1.1.1.log.txt).

**Comment by coryan** [comment]
The fixes for the documentation are in #3226

**Comment by coryan** [comment]
Closing as the questions on this seem to have been answered.
